README.md
[![Test Coverage](https://api.codeclimate.com/v1/badges/0139a26763ee4a0c3343/test_coverage)](https://codeclimate.com/github/averias/phpredis-json/test_coverage)
[![Maintainability](https://api.codeclimate.com/v1/badges/0139a26763ee4a0c3343/maintainability)](https://codeclimate.com/github/averias/phpredis-json/maintainability)
[![Build Status](https://travis-ci.org/averias/phpredis-json.svg?branch=master)](https://travis-ci.org/averias/phpredis-json)
[![Packagist Version](https://img.shields.io/packagist/v/averias/phpredis-json.svg)](https://packagist.org/packages/averias/phpredis-json)
[![GitHub](https://img.shields.io/github/license/averias/phpredis-json.svg)](https://github.com/averias/phpredis-json)

# Phpredis-JSON
RedisJson with the PHP Redis extension [phpredis](https://github.com/phpredis/phpredis).

## Intro
Phpredis-JSON provides a full set of commands for [RedisJson Module](https://oss.redislabs.com/redisjson/). 
It's built on top of the `phpredis` and use it as Redis client, 
so you can also take advantage of some of the features included in `phpredis` as Redis client.

## Why?
Although you can issue RedisJSON commands by using some PHP Redis clients which provides you the ability to send 
raw Redis commands, Phpredis-JSON:
- avoids you the task of encoding your PHP data structures to JSON before sending them to Redis 
and decode the JSON responses back from Redis
- it validates JSON encode/decode and throw a proper exception in case of failure
- provides a set of commands as methods of the RedisJSON client  

## Requirements
- Redis server 4.0+ version (RedisJson Module is only available from Redis 4.0+)
- RedisJson Module installed on Redis server as specified in [Building and loading RedisJSON Module](https://oss.redislabs.com/redisjson/#building-and-loading-the-module)
- PHP 7.2+ with PHP Redis extension 5 installed

## Usage
```php
<?php

use Averias\RedisJson\Enum\JsonCommands;
use Averias\RedisJson\Exception\ResponseException;
use Averias\RedisJson\Factory\RedisJsonClientFactory;

// *** get a RedisJsonClient ***
$redisJsonClientFactory = new RedisJsonClientFactory();
/**
 * creates a RedisJsonClient with default connection params:
 * [
 *     'host' => '127.0.0.1',
 *     'port' => 6379,
 *     'timeout' => 0.0, // seconds
 *     'retryInterval' => 15 // milliseconds
 *     'readTimeout' => 2, // seconds
 *     'persistenceId' => null // string for persistent connections, null for no persistent ones
 *     'database' => 0 // Redis database index [0..15]
 * ]
 */
$defaultClient = $redisJsonClientFactory->createClient();

// creates a configured RedisJsonClient
$client = $redisJsonClientFactory->createClient([
    'host' => '127.0.0.1',
    'port' => 6390,
    'timeout' => 2,
    'database' => 15
]);

// *** Redis JSON commands ***
$result = $client->jsonSet('people', ["name" => "gafael", "age" => 12]);
echo ($result === true ? 'true' : 'false') . PHP_EOL; // true

$result = $client->jsonGet('people'); // $result = ["name":"gafael","age":12]
echo json_encode($result) . PHP_EOL; // {"name":"gafael","age":12}

$result =  $client->jsonGet('people', '.name');
echo $result . PHP_EOL; // "gafael"

$result =  $client->jsonGet('people', '.age');
echo $result . PHP_EOL; // 12

// "nonexistent" key does not exist, so a ResponseException is thrown
try {
    $result = $client->jsonGet('nonexistent');
    echo $result . PHP_EOL;
} catch (ResponseException $e) {
    echo "key nonexistent does not exist" . PHP_EOL;
}

// *** you can also send RedisJSON command as raw commands using "RedisJsonClient::executeRawCommand"  ***
// you will send
$result =  $client->executeRawCommand(JsonCommands::SET, 'people', '.colors', '["blue", "green"]');
echo $result . PHP_EOL; // 'OK'

// and receive JSON values
$result =  $client->executeRawCommand(JsonCommands::GET, 'people', '.');
echo $result . PHP_EOL; // {"name":"gafael","age":12,"colors":["blue","green"]}


// *** you can also issue redis commands and use RedisJsonClient as "phpredis" client:
echo $client->hset('hash-key', 'age', 34) . PHP_EOL; // 0
echo $client->hget('hash-key', 'age') . PHP_EOL; // 34

// $ret = [true,"val1",true,"val2"]
$ret = $client->multi()
    ->set('key1', 'val1')
    ->get('key1')
    ->set('key2', 'val2')
    ->get('key2')
    ->exec();

echo json_encode($ret) . PHP_EOL;

```

## Commands
- **RedisJSON commands:** please take a look to the list of [phpredis-json commands](https://github.com/averias/phpredis-json/blob/master/docs/JSON-COMMANDS.md)
- **Phpredis commands:** you can send Redis commands as specified in [phpredis documentation](https://github.com/phpredis/phpredis#table-of-contents)
- **Raw commands:** you can send whatever you want to Redis by using `RedisJsonClient::executeRawCommand`:
```php
// raw Redis JSON command
$client->executeRawCommand(JsonCommands::GET, 'people', '.');

// raw Redis command
$client->executeRawCommand('hget, 'hash-key', 'foo');
``` 

## Tests
#### On a local Redis server 4.0+ with RedisJSON module and Redis extension 5 installed
From console run the following command from the root directory of this project:

`./vendor/bin/phpunit`

if you don't have configured your local Redis server in 127.0.0.1:6379 you can set REDIS_TEST_SERVER and REDIS_TEST_PORT 
and REDIS_TEST_DATABASE in `./phpunit.xml` file with your local Redis host, port and database before running the above 
command.
  
#### Docker
Having Docker installed, run the following command in the root directory of this project:

`bash run-tests-docker.sh`

by running the above bash script, two docker services will be built, one with PHP 7.2 with xdebug and redis extensions
enabled and another with the image of `redislab\rejson:1.0.4` (Redis server 5 with RedisJson module installed). 
Then the tests will run inside `phpredis-json` docker service container and finally both container will be stopped.

## Examples
- [Usage](https://github.com/averias/phpredis-json/blob/master/examples/usage.php)
- [Commands](https://github.com/averias/phpredis-json/blob/master/examples/commands.php)

## License
Phpredis-Json code is distributed under MIT license, see [LICENSE](https://github.com/averias/phpredis-json/blob/master/LICENSE) 
file


CHANGELOG.md
## 1.4.0 - 2021-02-27

- fixed wrong namespace for `Enum/Keys`

## 1.3.0 - 2020-01-24

- removed `symplify/changelog-linker` changelog generator
- changed Scrutinizer by CodeClimate

## 1.2.3 - 2020-01-24

- fixed wrong namespace for `JsonCommandTraitInterface`

## 1.2.2 - 2019-12-01

- added `symplify/changelog-linker` changelog generator

## 1.2.0/1 - 2019-11-26

- fixed issues in travis config
- removed deprecated setMethod in PhpUnit
- added scrutinizer
- added badgets to README.md

## 1.1.5 - 2019-11-23
- fixed typo in header documentation
- updated PhpUnit to version 8
- add travis ci and changed redis database from 0 to 15 for integration tests

## 1.1.4 - 2019-11-14

- fixed typo on header documentation
- delete commented code

## 1.1.3 - 2019-11-11

- removed invalid reference in phpunit.xml
- small cleaning and refactoring
- small changes in documentation

## 1.1.2 - 2019-11-10

- restored previous ways to get redis info because of a bug

## 1.1.1 - 2019-11-10

- fixed wrong namespaces

## 1.1.0 - 2019-11-09

- fixed typos in current documentation
- added php docker service and redisjson from redislab

## 1.0.1 - 2019-11-08

- finished documentation and examples

## 1.0.0 - 2019-11-07

- changed signature for Append String command and added some documents

## 0.1.x - 2019-11-07

- first implementation
- fixed typos in docs
- fixed name package in composer.json

examples/usage.php
<?php
/**
 * @project   phpredis-json
 * @author    Rafael Campoy <rafa.campoy@gmail.com>
 * @copyright 2019 Rafael Campoy <rafa.campoy@gmail.com>
 * @license   MIT
 * @link      https://github.com/averias/phpredis-json
 *
 * Copyright and license information, is included in
 * the LICENSE file that is distributed with this source code.
 */

namespace Examples;

require(dirname(__DIR__) . '/vendor/autoload.php');

use Averias\RedisJson\Client\RedisJsonClient;
use Averias\RedisJson\Enum\JsonCommands;
use Averias\RedisJson\Exception\ResponseException;
use Averias\RedisJson\Factory\RedisJsonClientFactory;

// instantiate factory
$redisJsonClientFactory = new RedisJsonClientFactory();

/**
 * creates a client with default connection params:
 * [
 *     'host' => '127.0.0.1',
 *     'port' => 6379,
 *     'timeout' => 0.0, // seconds
 *     'retryInterval' => 15 // milliseconds
 *     'readTimeout' => 2, // seconds
 *     'persistenceId' => null // string for persistent connections, null for no persistent ones
 *     'database' => 0 // Redis database index [0..15]
 * ]
 */

/** @var RedisJsonClient $client */
$client = $redisJsonClientFactory->createClient();

// creates a configured client
$client = $redisJsonClientFactory->createClient([
    'host' => '127.0.0.1',
    'port' => 6379,
    'timeout' => 2,
    'database' => 15
]);

/**
 * creates a RedisJsonClient with default connection params:
 * [
 *     'host' => '127.0.0.1',
 *     'port' => 6379,
 *     'timeout' => 0.0, // seconds
 *     'retryInterval' => 15 // milliseconds
 *     'readTimeout' => 2, // seconds
 *     'persistenceId' => null // string for persistent connections, null for no persistent ones
 *     'database' => 0 // Redis database index [0..15]
 * ]
 */

// json commands
$result = $client->jsonSet('people', ["name" => "gafael", "age" => 12]);
echo ($result === true ? 'true' : 'false') . PHP_EOL; // true

$result = $client->jsonGet('people'); // $result = ["name":"gafael","age":12]
echo json_encode($result) . PHP_EOL; // {"name":"gafael","age":12}

$result = $client->jsonGet('people', '.name');
echo $result . PHP_EOL; // "gafael"

$result = $client->jsonGet('people', '.age');
echo $result . PHP_EOL; // 12

// "nonexistent" key does not exists, so a ResponseException is thrown
try {
    $result = $client->jsonGet('nonexistent');
    echo $result . PHP_EOL;
} catch (ResponseException $e) {
    echo "key nonexistent does not exist" . PHP_EOL;
}

// you can also send RedisJSON command as raw commands using "executeRawCommand", you will send a receive JSON values
$result = $client->executeRawCommand(JsonCommands::SET, 'people', '.colors', '["blue", "green"]');
echo $result . PHP_EOL; // 'OK'

$result = $client->executeRawCommand(JsonCommands::GET, 'people', '.');
echo $result . PHP_EOL; // {"name":"gafael","age":12,"colors":["blue","green"]}


// you can also issue redis commands and use RedisJsonClient as "phpredis" client:
echo $client->hset('hash-key', 'age', 34) . PHP_EOL; // 0
echo $client->hget('hash-key', 'age') . PHP_EOL; // 34

// $ret = [true,"val1",true,"val2"]
$ret = $client->multi()
    ->set('key1', 'val1')
    ->get('key1')
    ->set('key2', 'val2')
    ->get('key2')
    ->exec();

echo json_encode($ret) . PHP_EOL;
